### PR TITLE
Bump typing-extensions and typeapi

### PR DESCRIPTION
**Pull request by rhaps0dy**
`typing-extensions<4.7` conflicts with `torch==2.2.1`, and is only necessary for python 3.7 or (probably) 3.6. This PR makes the maximum `typing-extensions` version larger for releases after 3.7, and also extends the range for typeapi.

### CODE REVIEW

**Comment by rhaps0dy** [comment]
I think we should make this version 4.5.0 .

**Comment by NiklasRosenstein** [comment]
LGTM! Only Mypy has recently release a version that no longer supports Python 3.6. I'm happy to throw out 3.6/3.7 in Databind as well. I've actually just released a new `typeapi` version ([2.2.0](https://pypi.org/project/typeapi/2.2.0/)) that dropped support for these versions.

**Comment by rhaps0dy** [comment]
I've removed code and annotations that supported python 3.6 and 3.7 as best as I could. However there are still some things I don't know how to remove correctly (or whether to in the first place), so I would appreciate a second look:

```text
% rg 3\.7
% rg 3\.6
databind.core/src/databind/core/schema.py
226:                # NOTE(NiklasRosenstein): In Python 3.6, Mypy complains about "Callable does not accept self argument",

databind.core/src/databind/core/settings.py
475:        return id(self)  # Needs to be hashable for Annotated[...] in Python 3.6

docs/content/core/settings.md
35:To support Python 3.6, `typing_extensions.Annotated` metadata must be hashable. As such, settings that you want users
36:to be able to pass into annotated type hints in Python 3.6 must support hashing (identity hashing is often sufficient).
```

File: databind.json/src/databind/json/converters.py [hunk 1]
```diff
@@ -167,7 +166,6 @@ def _length_check() -> None:
             try:
                 return python_type(values)
             except TypeError:
-                assert not isinstance(values, types.GeneratorType), (type(values), python_type)
```
Comment (rhaps0dy) [inline databind.json/src/databind/json/converters.py:170]: This gave an Unreachable error in mypy, probably it's due to the newest mypy version knowing to check this

**Pull request merged**